File: COMP2041/lec/02_Unix_Filters/sort.py
#!/usr/bin/python3

# Simple sort emulation for a COMP2041/9044 lecture example
# andrewt@unsw.edu.au

import sys


def process_stream(f):
    """
    print lines of stream in sorted order
    """
    print("".join(sorted(f)), end="")


def main():
    """
    process files given as arguments, if no arguments process stdin
    """
    if len(sys.argv) == 1:
        process_stream(sys.stdin)
    else:
        with open(sys.argv[1], 'r') as f:
            process_stream(f)


if __name__ == '__main__':
    main()

File: COMP2041/lec/02_Unix_Filters/sort.c
// Simple sort emulation for a COMP2041/9044 lecture example
// andrewt@unsw.edu.au

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int cmpstringp(const void *p1, const void *p2)
{
   return strcmp(* (char * const *) p1, * (char * const *) p2);
}


// print lines of stream in sorted order
void process_stream(FILE *stream) {
    char *line = NULL;
    size_t line_size = 0;
    char **lines = NULL;
    size_t lines_size = 0;

    while (getline(&line, &line_size, stream) > 0) {
        lines = realloc(lines, ++lines_size*sizeof(lines[0]));
        lines[lines_size-1] = line;
        line = NULL;
        line_size = 0;
    }
    free(line);

    qsort(lines, lines_size, sizeof(lines[0]), cmpstringp);

    for (size_t i = 0; i < lines_size; i++) {
        fputs(lines[i], stdout);
        free(lines[i]);
    }
    free(lines);
}

// process files given as arguments
// if no arguments process stdin
int main(int argc, char *argv[]) {

    if (argc == 1) {
        process_stream(stdin);
    } else {
        for (int i = 1; i < argc; i++) {
            FILE *in = fopen(argv[i], "r");
            if (in == NULL) {
                fprintf(stderr, "%s: %s: ", argv[0], argv[i]);
                perror("");
                return 1;
            }
            process_stream(in);
            fclose(in);
        }
    }

    return 0;
}
